ir/src/constraint_builder/variables.rs

```diff
@@ -1,73 +1,174 @@
+                        ident.name(),
+                        &access_type,
+                    ));
+                }
+            };
+            // reduce the dimension of the access by 2, since we indexed the bound value twice.
+            let inner_access_type = reduce_access_dim(ident.name(), access_type)?;
+            let inner_access_type = reduce_access_dim(ident.name(), inner_access_type)?;
+            (inner_expr, inner_access_type)
         }
         _ => {
             return Err(SemanticError::invalid_variable_access_type(
-                variable_name,
-                access_type,
-            ))
+                ident.name(),
+                &access_type,
+            ));
         }
     };
 
-    Ok(expr)
+    // access the inner expression with the specified access type to get the expression
+    access_inner_expr(ident.name(), inner_expr, inner_access_type)
+}
+
+// HELPERS
+// ================================================================================================
+
+/// Returns a new [AccessType] with the dimension reduced by one. For example, a Matrix access
+/// becomes a Vector access.
+fn reduce_access_dim(var_name: &str, access_type: AccessType) -> Result<AccessType, SemanticError> {
+    match access_type {
+        AccessType::Default => Err(SemanticError::invalid_variable_access_type(
+            var_name,
+            &access_type,
+        )),
+        AccessType::Vector(_) => Ok(AccessType::Default),
+        AccessType::Matrix(_, idx) => Ok(AccessType::Vector(idx)),
+    }
+}
+
+/// Accesses into a `BindingAccess` expression and returns a new `BindingAccess` of a higher
+/// dimension.
+///
+/// For example:
+/// - suppose the `expr` is a [BindingAccess] specifying that a binding `A` is being accessed as a
+///   vector at index 0 (i.e. the access_type is [AccessType::Vector(0)], representing `A[0]`),
+/// - and suppose the specified `access_type` is [AccessType::Vector(i)].
+/// - Then the resulting expression would be A[i][0], represented by a new [BindingAccess] with
+///   identifier `A` and [AccessType::Matrix(0, i)].
+///
+/// # Errors
+/// Returns an error if the expression is one that can't be accessed
+fn access_inner_expr(
+    var_name: &str,
+    expr: &Expression,
+    access_type: AccessType,
+) -> Result<Expression, SemanticError> {
+    match access_type {
+        // access the entire expression
+        AccessType::Default => Ok(expr.clone()),
+        // access into the expression at the specified index
+        AccessType::Vector(new_idx) => match expr {
+            Expression::BindingAccess(inner_binding) => match inner_binding.access_type() {
+                AccessType::Default => {
+                    let new_binding_access = BindingAccess::new(
+                        inner_binding.ident().clone(),
+                        AccessType::Vector(new_idx),
+                    );
+                    Ok(Expression::BindingAccess(new_binding_access))
+                }
```

Comment:
I got lost in this a little. Does this mean that if we have something like:
```
let x = [1, 2, 3]
let c = x[2]
```
We replace `c = x[2]` with `c = 3`?

Reply:
No, it means we replace `c` where it is used by `x[2]`. In other words, if we had something like the following:

```
let x = [1, 2, 3]
let c = x[2]
enf c = 3
```

this step  turns it into:

```
let x = [1, 2, 3]
let c = x[2]
enf x[2] = 3
```

Then a subsequent processing turns it into `enf 3 = 3`